TODO: `core.algorithms.simplfy.loop()` -- needs to be vectorized

Start of `core.algorithms.simplfy.loop()`

```python
    dropped = edges[es_mask].geometry.item()
    segments = list(
        map(
            shapely.LineString,
            zip(dropped.coords[:-1], dropped.coords[1:], strict=True),
        )
    )  # TODO: vectorize this
```

xref uscuni/simplification#95

#### Amendment 2

Swapping out the exact same function in [`artifacts.n1_g1_identical()`](https://github.com/uscuni/sgeop/blob/b5db0feb007f213438b31af218f574f8a6285b1b/sgeop/artifacts.py#L720-L725) leads to 2 failures and 8 warnings.

```
============================================================================= warnings summary ==============================================================================
sgeop/tests/test_simplify.py::test_simplify_network_full_fua[auckland_869-0.3-1268048]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (298999.6896515278 -4076075.8698911807). The artifact has not been simplified. The original message:
  too many indices for array: array is 1-dimensional, but 2 were indexed
    roads = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[auckland_869-0.3-1268048]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (309435.29298630304 -4086815.291779113). The artifact has not been simplified. The original message:
  too many indices for array: array is 1-dimensional, but 2 were indexed
    roads = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[liege_1656-0.3-2350782]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (675045.8652641856 5612681.851142581). The artifact has not been simplified. The original message:
  IllegalArgumentException: point array must contain 0 or >1 elements
  
    roads = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[liege_1656-0.3-2350782]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (675688.4388394251 5610464.265210363). The artifact has not been simplified. The original message:
  too many indices for array: array is 1-dimensional, but 2 were indexed
    roads = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[liege_1656-0.3-2350782]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (684968.3704674016 5616998.1691926). The artifact has not been simplified. The original message:
  too many indices for array: array is 1-dimensional, but 2 were indexed
    roads = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[liege_1656-0.3-2350782]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (685382.1750726631 5610738.626389066). The artifact has not been simplified. The original message:
  too many indices for array: array is 1-dimensional, but 2 were indexed
    roads = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[slc_4881-0.3-1762456]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:355: UserWarning: An error occured at location POINT (428886.8113532481 4512617.5603151). The artifact has not been simplified. The original message:
  too many indices for array: array is 1-dimensional, but 2 were indexed
    roads_cleaned = simplify_singletons(

sgeop/tests/test_simplify.py::test_simplify_network_full_fua[aleppo_1133-0.2-4361625]
  /Users/jgaboardi/github_repos/uscuni/sgeop/sgeop/simplify.py:732: UserWarning: An error occured at location POINT (336049.0533133858 4006921.702477669). The artifact has not been simplified. The original message:
  IllegalArgumentException: point array must contain 0 or >1 elements
  
    roads = simplify_singletons(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform darwin, python 3.12.7-final-0 ----------
Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
sgeop/__init__.py         8      0   100%
sgeop/artifacts.py      376     12    97%   95, 192-196, 204, 648-653, 1213-1215
sgeop/continuity.py      40      0   100%
sgeop/geometry.py       143      0   100%
sgeop/nodes.py          189      0   100%
sgeop/simplify.py       182      0   100%
---------------------------------------------------
TOTAL                   938     12    99%

========================================================================== short test summary info ==========================================================================
FAILED sgeop/tests/test_simplify.py::test_simplify_network_small[standard-1.5-64566.0] - assert 64576.111663542615 ± 6.5e+00 == 64566.0
FAILED sgeop/tests/test_simplify.py::test_simplify_network_small[exclusion_mask-1.05-65765.0] - assert 65775.78242082833 ± 6.6e+00 == 65765.0
=========================================================== 2 failed, 163 passed, 8 warnings in 216.23s (0:03:36) ===========================================================
```

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -600,12 +600,15 @@ def loop(
     # check if we need to add a deadend to represent the space
     to_add = []
     dropped = edges[es_mask].geometry.item()
+    """
     segments = list(
         map(
             shapely.LineString,
             zip(dropped.coords[:-1], dropped.coords[1:], strict=True),
         )
     )  # TODO: vectorize this
+    """
+    segments = shapely.linestrings((dropped.coords[:-1], dropped.coords[1:]))
 
     # figure out if there's a snapping node
     # Get nodes on Cs
@@ -717,12 +720,15 @@ def n1_g1_identical(
     to_drop.append(edges.index[0])
     dropped = edges.geometry.item()
 
+    """
     segments = list(
         map(
             shapely.LineString,
             zip(dropped.coords[:-1], dropped.coords[1:], strict=True),
         )
     )
+    """
+    segments = shapely.linestrings((dropped.coords[:-1], dropped.coords[1:]))
 
     snap_to = shapely.get_point(dropped, 0)
 

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -1413,8 +1413,5 @@ def is_dangle(edgelines):
 
 def line_segments(line: shapely.LineString) -> np.ndarray:
     """Explode a linestring into constituent pairwise coordinates."""
-    return shapely.linestrings(
-        np.c_[np.array(line.coords)[:-1], np.array(line.coords)[1:]].reshape(
-            np.array(line.coords).shape[0] - 1, 2, 2
-        )
-    )
+    xys = np.array(line.coords)
+    return shapely.linestrings(np.c_[xys[:-1], xys[1:]].reshape(xys.shape[0] - 1, 2, 2))

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -1413,7 +1413,8 @@ def is_dangle(edgelines):
 
 def line_segments(line: shapely.LineString) -> np.ndarray:
     """Explode a linestring into constituent pairwise coordinates."""
-    xys = line.coords[:]
     return shapely.linestrings(
-        np.c_[np.array(xys[:-1]), np.array(xys[1:])].reshape(len(xys) - 1, 2, 2)
+        np.c_[np.array(line.coords)[:-1], np.array(line.coords)[1:]].reshape(
+            np.array(line.coords).shape[0] - 1, 2, 2
+        )
     )

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -600,15 +600,7 @@ def loop(
     # check if we need to add a deadend to represent the space
     to_add = []
     dropped = edges[es_mask].geometry.item()
-    """
-    segments = list(
-        map(
-            shapely.LineString,
-            zip(dropped.coords[:-1], dropped.coords[1:], strict=True),
-        )
-    )  # TODO: vectorize this
-    """
-    segments = shapely.linestrings((dropped.coords[:-1], dropped.coords[1:]))
+    segments = line_segments(dropped)
 
     # figure out if there's a snapping node
     # Get nodes on Cs
@@ -719,16 +711,7 @@ def n1_g1_identical(
 
     to_drop.append(edges.index[0])
     dropped = edges.geometry.item()
-
-    """
-    segments = list(
-        map(
-            shapely.LineString,
-            zip(dropped.coords[:-1], dropped.coords[1:], strict=True),
-        )
-    )
-    """
-    segments = shapely.linestrings((dropped.coords[:-1], dropped.coords[1:]))
+    segments = line_segments(dropped)
 
     snap_to = shapely.get_point(dropped, 0)
 
@@ -1426,3 +1409,11 @@ def is_dangle(edgelines):
     ).sum(axis=1)
 
     return (first_sum == 1) | (last_sum == 1)
+
+
+def line_segments(line: shapely.LineString) -> np.ndarray:
+    """Explode a linestring into constituent pairwise coordinates."""
+    xys = line.coords[:]
+    return shapely.linestrings(
+        np.c_[np.array(xys[:-1]), np.array(xys[1:])].reshape(len(xys) - 1, 2, 2)
+    )

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -1408,7 +1408,7 @@ def is_dangle(edgelines):
 
 def line_segments(line: shapely.LineString) -> np.ndarray:
     """Explode a linestring into constituent pairwise coordinates."""
-    xys = np.array(line.coords)
+    xys = shapely.get_coordinates(line)
     return shapely.linestrings(
         np.column_stack((xys[:-1], xys[1:])).reshape(xys.shape[0] - 1, 2, 2)
     )

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -1303,12 +1303,7 @@ def nx_gx_cluster(
             # this is a fallback for corner cases. It should result in the nearly the
             # same skeleton in the end but ensures we work with a single-part geometry
             merged_edges = shapely.concave_hull(merged_edges).exterior
-        skeletonization_input = list(
-            map(
-                shapely.LineString,
-                zip(merged_edges.coords[:-1], merged_edges.coords[1:], strict=True),
-            )
-        )
+        skeletonization_input = line_segments(merged_edges)
 
     # skeletonize
     skel, _ = voronoi_skeleton(

--- a/sgeop/artifacts.py
+++ b/sgeop/artifacts.py
@@ -1414,4 +1414,6 @@ def is_dangle(edgelines):
 def line_segments(line: shapely.LineString) -> np.ndarray:
     """Explode a linestring into constituent pairwise coordinates."""
     xys = np.array(line.coords)
-    return shapely.linestrings(np.c_[xys[:-1], xys[1:]].reshape(xys.shape[0] - 1, 2, 2))
+    return shapely.linestrings(
+        np.column_stack((xys[:-1], xys[1:])).reshape(xys.shape[0] - 1, 2, 2)
+    )

